### PR TITLE
fix: collect cached/reasoning token counts from OpenAI usage

### LINKED ISSUE #115: OpenAI Responses usage drops cached & reasoning token counts (Copilot gpt-5.5)
## Summary
When using the OpenAI **Responses API** — including through **GitHub Copilot** (e.g. `gpt-5.5`) — prompt-cache and reasoning token counts are never collected. `Usage.TotalCachedTokens` and `Usage.TotalReasoningTokens` always read `0`, even on large cache hits, so cache/cost telemetry is wrong.

## Steps to Reproduce
1. Use a Responses-API model (e.g. `gpt-5.5` via Copilot) in `LmStreaming.Sample`.
2. Send a prompt large enough to trigger prompt caching (or replay a recorded `response.completed` event that carries a non-zero `cached_tokens`).
3. Observe the emitted `UsageMessage` / the sample's `Cache:` log line.

## Expected Behavior
The cached input tokens (and reasoning tokens) reported by the API are surfaced on the `UsageMessage` — `TotalCachedTokens` / `TotalReasoningTokens` reflect the real counts.

## Actual Behavior
`TotalCachedTokens` / `TotalReasoningTokens` are always `0`; the sample logs `Cache: read=0` for every Responses-API turn.

## Root Cause
`src/OpenAiResponsesProvider/Agents/OpenAiResponsesAgent.cs` → `TryExtractUsage` read only the three top-level fields (`input_tokens`, `output_tokens`, `total_tokens`) and **ignored the nested detail objects**:

- `input_tokens_details.cached_tokens` → dropped → `TotalCachedTokens == 0`
- `output_tokens_details.reasoning_tokens` → dropped → `TotalReasoningTokens == 0`

The core `Usage` model (`src/LmCore/Models/Usage.cs`) already supports both. The GitHub Copilot path delegates to the same `OpenAiResponsesAgent` (`CopilotResponsesAgentFactory`), and both the SSE and WebSocket transports share the agent's event→message mapping, so the single parser was the only defect for the Responses surface.

A secondary reason this went unnoticed: the test mock (`OpenAiResponsesEventStreamWriter.BuildResponseSnapshot`) also only emitted the three top-level fields, so no test could ever exercise the nested-detail path.

## Repro Proof
A **real captured** `response.completed` event (`samples/MockProviderHost/fixtures/openai-responses-websocket/sample_ws_stream_001.redacted.json`) carries:

```json
{
  "input_tokens": 14986,
  "input_tokens_details": { "cached_tokens": 13696 },
  "output_tokens": 121,
  "output_tokens_details": { "reasoning_tokens": 0 },
  "total_tokens": 15107
}
```

13,696 of 14,986 input tokens (~91%) were served from cache — upstream **does** send it; we dropped it. A regression test (`tests/OpenAiResponsesProvider.Tests/CacheTokenUsageRegressionTests.cs`) replays this event through the agent and asserts `TotalCachedTokens`:

- **Before fix (RED):** `Expected usage.Usage.TotalCachedTokens to be 13696 … but found 0`
- **After fix (GREEN):** passes

## Scope / Related Risk
The same "rebuild a core `Usage` by hand and drop the nested details" pattern exists on other provider paths (OpenAI/OpenRouter chat-completions, non-streaming Anthropic, OpenRouter merge middleware) and there is a cross-provider `PromptTokens` semantic inconsistency. The chat-completions emit path is fixed alongside this Responses fix in the linked PR; the remaining items are tracked in a separate follow-up issue.

### PR DESCRIPTION
**Pull request by gautam-achieveai**
## What & Why
Prompt-cache and reasoning token counts were silently dropped for OpenAI usage, so `Usage.TotalCachedTokens` / `TotalReasoningTokens` always read `0` — even on large cache hits (a real captured Responses event showed **13696/14986 input tokens cached, ~91%**, all discarded).

The OpenAI **Responses API** (including GitHub Copilot `gpt-5.5`) and **Chat Completions** report cache/reasoning in *nested* usage detail objects that the providers weren't reading. This is the bug reported for "OpenAI responses APIs (through copilot)" (#115); the chat-completions half of the same family is fixed here too (opted in), with the rest tracked in #116.

## Root Cause
- **Responses:** `OpenAiResponsesAgent.TryExtractUsage` read only `input_tokens` / `output_tokens` / `total_tokens` and ignored `input_tokens_details.cached_tokens` + `output_tokens_details.reasoning_tokens`.
- **Chat Completions:** `OpenAIProviderUsage` only mapped the *Responses-API* detail names (`input_tokens_details`), not the chat-completions names (`prompt_tokens_details` / `completion_tokens_details`), **and** the emit path (`OpenUsage` → `OpenMessage`) rebuilt a bare `Usage` that dropped details regardless.

## Changes
- **`OpenAiResponsesAgent.TryExtractUsage`** — parse the nested `input_tokens_details` / `output_tokens_details` into `Usage.InputTokenDetails` / `OutputTokenDetails`. Covers SSE + WebSocket, direct + Copilot (shared agent).
- **`OpenAIProviderUsage`** — also parse the chat-completions `prompt_tokens_details` / `completion_tokens_details` names; `ToCoreUsage` now maps details via the unified `TotalCachedTokens` / `TotalReasoningTokens` so OpenRouter-direct fields reach core too.
- **`OpenUsage` / `OpenMessage` / `OpenClientAgent`** — carry cached/reasoning onto the emitted `UsageMessage` (non-streaming + streaming).
- **`OpenAiResponsesEventStreamWriter` (mock)** — emit the nested details when a plan simulates cache/reasoning, closing the test-infra gap that hid this.
- **`LmStreaming.Sample` `Cache:` log** — report true uncached input (`prompt - cached` for the OpenAI family; defensive fallback for the Anthropic additive case; clamped ≥ 0).

## Tests (RED→GREEN)
- `OpenAiResponsesProvider.Tests/CacheTokenUsageRegressionTests` — replays the real `13696`-cached `response.completed` event through the agent (failed `found 0` before, passes after).
- `OpenAiResponsesProvider.Tests/OpenAiResponsesEventStreamWriterTests` — mock emits / omits nested details.
- `OpenAIProvider.Tests/Agents/ChatCompletionsCacheTokenUsageTests` — end-to-end through `OpenClientAgent` with a canned chat-completion body.
- `OpenAIProvider.Tests/Models/OpenAIProviderUsageTests` — chat-completions nested-detail parsing + `ToCoreUsage`.

All green: OpenAiResponsesProvider 70 · MockProviderHost 41 · LmTestUtils 91 · OpenAIProvider 91 · LmStreaming.Sample 325 · LmCore (Usage) 16 · GithubCopilotProvider 19. Full solution build: 0 errors.

## Risk
Additive parsing + emit changes; defaults preserve prior behavior when no details are present (existing byte-identity and usage tests unchanged). The sample-log change is diagnostic only.

## Related Issues
Fixes #115
Relates to #116

### CODE REVIEW

**Review by gautam-achieveai — COMMENTED** [review]
**[Gautam's review bot]** :white_check_mark: **APPROVE**

**Bug context:** #115 — OpenAI Responses usage drops cached & reasoning token counts (Copilot gpt-5.5). The nested `*_tokens_details` objects were being silently dropped across both the Responses API and Chat Completions API paths.

**Review summary:**
- Clean, surgical fix that correctly addresses the bug across both provider paths (Responses + Chat Completions)
- Excellent test coverage: RED→GREEN regression tests with real captured 13696-cached fixture data
- `TryReadNestedInt` helper is clean and defensive
- `BuildCoreUsage` eliminates previous code duplication between streaming and non-streaming paths
- Unified accessor approach in `ToCoreUsage()` simplifies all three shapes (Responses-API nesting, Chat Completions nesting, OpenRouter direct) into one consistent mapping
- The mock infrastructure (`OpenAiResponsesEventStreamWriter`) is properly updated to exercise the parsing paths

**Minor observation (non-blocking):**
The legacy `OpenUsage.IsCached` (bool) is now somewhat redundant alongside the new `CachedTokens` (int) field. Worth a follow-up cleanup but not blocking.

All 22 relevant tests pass (11 OpenAI provider + 11 Responses provider).

**Review by gautam-achieveai — COMMENTED** [review]
**[Codex's bot]** **APPROVE WITH COMMENTS**

Reviewed #117 with parallel specialist passes (testing, schema compatibility, performance, architecture, duplication, simplification, EUII/logging, scope/over-engineering, feature-flag risk, exception handling, and temp/security). The change is directionally aligned with #115 and I did not find any blocking issue.

Findings posted:
- **2 MEDIUM**: add streaming Chat Completions coverage; make optional Responses usage-detail parsing non-throwing.
- **2 LOW**: centralize repeated token-detail mapping; avoid provider-specific cache normalization in the WebSocket/sample layer.
- **1 QUESTION**: clarify the core `Usage.InputTokenDetails.CachedTokens` contract.

Verification note: `git diff --check origin/main...HEAD` passed. I could not run the targeted .NET tests in this sandbox because `dotnet` is not installed (`dotnet: not found`).

File: tests/OpenAIProvider.Tests/Agents/ChatCompletionsCacheTokenUsageTests.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        """;
+
+    [Fact]
+    public async Task UsageMessage_preserves_cached_and_reasoning_tokens_from_chat_completions()
+    {
+        using var httpClient = new HttpClient(FakeHttpMessageHandler.CreateSimpleJsonHandler(ResponseJson));
+        var client = new OpenClient(httpClient, BaseUrl);
+        var agent = new OpenClientAgent("TestAgent", client);
+
+        var response = await agent.GenerateReplyAsync(
```
Comment (gautam-achieveai) [inline tests/OpenAIProvider.Tests/Agents/ChatCompletionsCacheTokenUsageTests.cs:48]: **[Codex's bot]** **MEDIUM** (Testing)

This regression test only exercises the non-streaming `GenerateReplyAsync` path, but the PR also changes the streaming Chat Completions usage propagation (`OpenAgent` now copies `TotalCachedTokens` / `TotalReasoningTokens` into streamed `OpenUsage`). That leaves the SSE parsing + `OpenMessage.ToStreamingMessage()` path unprotected, including the conditional that only emits a `UsageMessage` for non-zero streamed usage.

**Suggestion:** Add a streaming `OpenClientAgent.GenerateReplyStreamingAsync(...)` regression test with a final usage chunk containing:

```json
"usage": {
  "prompt_tokens": 14986,
  "completion_tokens": 121,
  "total_tokens": 15107,
  "prompt_tokens_details": { "cached_tokens": 13696 },
  "completion_tokens_details": { "reasoning_tokens": 64 }
}
```

Collect the streamed `UsageMessage` and assert `TotalCachedTokens == 13696` and `TotalReasoningTokens == 64`.

File: src/OpenAiResponsesProvider/Agents/OpenAiResponsesAgent.cs [hunk 2]
```diff
@@ -483,15 +483,34 @@ private static bool TryExtractUsage(JsonElement responseElement, out Usage usage
...
         return true;
     }
 
+    private static int TryReadNestedInt(JsonElement parent, string objectProperty, string numberProperty)
+    {
+        return parent.TryGetProperty(objectProperty, out var nestedEl)
+            && nestedEl.ValueKind == JsonValueKind.Object
+            && nestedEl.TryGetProperty(numberProperty, out var valueEl)
+            && valueEl.ValueKind == JsonValueKind.Number
+            ? valueEl.GetInt32()
```
Comment (gautam-achieveai) [inline src/OpenAiResponsesProvider/Agents/OpenAiResponsesAgent.cs]: **[Codex's bot]** **MEDIUM** (Exception Handling / Robustness)

`TryReadNestedInt` checks `ValueKind == Number` and then calls `GetInt32()`. `GetInt32()` can still throw for valid upstream numeric variation (for example an out-of-range count or non-int numeric encoding). Since this is optional usage telemetry parsed during `response.completed`, one unexpected detail value can abort an otherwise valid stream instead of just dropping the optional metric.

**Suggestion:** Make the helper non-throwing with `TryGetInt32` (or `TryGetInt64` plus an explicit clamp/ignore policy):

```csharp
private static int TryReadNestedInt(JsonElement parent, string objectProperty, string numberProperty)
{
    return parent.TryGetProperty(objectProperty, out var nestedEl)
        && nestedEl.ValueKind == JsonValueKind.Object
        && nestedEl.TryGetProperty(numberProperty, out var valueEl)
        && valueEl.ValueKind == JsonValueKind.Number
        && valueEl.TryGetInt32(out var value)
            ? value
            : 0;
}
```

File: src/OpenAIProvider/Models/OpenAIProviderUsage.cs [hunk 3]
```diff
@@ -78,21 +89,16 @@ public Usage ToCoreUsage()
...
-        if (InputTokenDetails != null)
+        // Convert nested token details using the unified accessors so every shape (Responses-API
+        // nesting, Chat Completions nesting, and OpenRouter direct fields) maps to core consistently.
+        if (TotalCachedTokens > 0)
         {
-            usage = usage with
-            {
-                InputTokenDetails = new InputTokenDetails { CachedTokens = InputTokenDetails.CachedTokens },
-            };
+            usage = usage with { InputTokenDetails = new InputTokenDetails { CachedTokens = TotalCachedTokens } };
```
Comment (gautam-achieveai) [inline src/OpenAIProvider/Models/OpenAIProviderUsage.cs]: **[Codex's bot]** **LOW** (Maintainability / Duplication)

This PR now has the same cached/reasoning-token → core `Usage.InputTokenDetails` / `OutputTokenDetails` mapping in several places (`OpenAIProviderUsage.ToCoreUsage`, `OpenMessage.BuildCoreUsage`, and `OpenAiResponsesAgent.TryExtractUsage`). That is easy to drift the next time another provider usage field is added.

**Suggestion:** Centralize the post-parse mapping in one helper/factory, e.g. `Usage.FromTokenCounts(...)`, `Usage.WithTokenDetails(...)`, or `OpenUsage.ToCoreUsage()`, then reuse it from both provider paths.

File: samples/LmStreaming.Sample/WebSocket/ChatWebSocketManager.cs [hunk 4]
```diff
@@ -224,10 +224,19 @@ private async Task StreamMessagesToClientAsync(
                 {
                     var u = usageMsg.Usage;
                     var cacheCreation = u.GetExtraProperty<int>("cache_creation_input_tokens");
+                    // cached_tokens is a SUBSET of PromptTokens for the OpenAI family (Responses + chat
```
Comment (gautam-achieveai) [inline samples/LmStreaming.Sample/WebSocket/ChatWebSocketManager.cs:227]: **[Codex's bot]** **LOW** (Architecture / Provider Semantics)

The WebSocket/sample logging layer now embeds provider-specific token-accounting semantics (OpenAI cached tokens as a subset of prompt tokens vs Anthropic cache reads being additive) to compute `uncached_input`. That normalization does not really belong in the presentation/logging layer and can drift as provider semantics evolve.

**Suggestion:** Move this behind a core `Usage` helper/normalization API (or log raw prompt/cache values only until #116 settles the cross-provider contract). The WebSocket layer should not need to know provider-specific cache accounting rules.

File: samples/LmStreaming.Sample/WebSocket/ChatWebSocketManager.cs [hunk 5]
```diff
@@ -224,10 +224,19 @@ private async Task StreamMessagesToClientAsync(
...
+                    // completions), so uncached = prompt - cached. Anthropic instead reports cache reads
+                    // SEPARATELY from input_tokens, so when the cache read exceeds the reported prompt we
+                    // fall back to PromptTokens (the additive case) and never go negative. The proper
+                    // cross-provider normalization of Usage is tracked as a follow-up.
+                    var uncachedInput = u.TotalCachedTokens <= u.PromptTokens
```
Comment (gautam-achieveai) [inline samples/LmStreaming.Sample/WebSocket/ChatWebSocketManager.cs:232]: **[Codex's bot]** [QUESTION] **Clarification Needed**

**Code:**
`var uncachedInput = u.TotalCachedTokens <= u.PromptTokens ? u.PromptTokens - u.TotalCachedTokens : u.PromptTokens;`

**Question:** What is the intended core contract for `Usage.InputTokenDetails.CachedTokens`: is it always a subset of `PromptTokens`, or can it represent additional cache-read tokens separate from `PromptTokens` depending on provider?

**Why this matters:** Answering this determines whether consumers should subtract cached tokens themselves, whether the core model needs an explicit normalized uncached-input helper, and how #116 should define cross-provider usage semantics.

**Possible answers:**
- It is always normalized as a subset of `PromptTokens`; providers should adapt before populating core `Usage`.
- It preserves provider-native semantics; consumers must inspect provider/follow a helper.
- The contract is not settled yet; this sample should avoid derived `uncached_input` until #116.

File: tests/OpenAIProvider.Tests/Agents/ChatCompletionsCacheTokenUsageTests.cs (hunk 1 above)
Comment (gautam-achieveai) [reply tests/OpenAIProvider.Tests/Agents/ChatCompletionsCacheTokenUsageTests.cs:48]: Addressed in eda44db — added `Streaming_UsageMessage_preserves_cached_and_reasoning_tokens_from_chat_completions` in this file. It drives `OpenClientAgent.GenerateReplyStreamingAsync` with a final usage chunk carrying `prompt_tokens_details` / `completion_tokens_details` and asserts `TotalCachedTokens == 13696` / `TotalReasoningTokens == 64` on the streamed `UsageMessage`, exercising `OpenMessage.ToStreamingMessage()` including the non-zero-usage gate. (The streaming reader requires each chunk to carry a choice/delta, so usage is co-located on the final stop chunk rather than an empty-choices chunk.)

File: src/OpenAiResponsesProvider/Agents/OpenAiResponsesAgent.cs (hunk 2 above)
Comment (gautam-achieveai) [reply src/OpenAiResponsesProvider/Agents/OpenAiResponsesAgent.cs]: Fixed in eda44db — `TryReadNestedInt` now uses `TryGetInt32` (your suggested form), so an out-of-range or non-int numeric in optional usage telemetry returns 0 instead of throwing and aborting an otherwise valid stream.

File: src/OpenAIProvider/Models/OpenAIProviderUsage.cs (hunk 3 above)
Comment (gautam-achieveai) [reply src/OpenAIProvider/Models/OpenAIProviderUsage.cs]: Done in eda44db — centralized the mapping in `Usage.WithTokenDetails(cachedTokens, reasoningTokens)` (LmCore) and reused it from `OpenAiResponsesAgent.TryExtractUsage`, `OpenMessage.BuildCoreUsage`, and `OpenAIProviderUsage.ToCoreUsage`. The 3-way duplication is gone.

File: samples/LmStreaming.Sample/WebSocket/ChatWebSocketManager.cs (hunk 4 above)
Comment (gautam-achieveai) [reply samples/LmStreaming.Sample/WebSocket/ChatWebSocketManager.cs:227]: Kept the derivation but contained it: the sample-layer guard (`cached <= prompt ? prompt - cached : prompt`) is intentionally minimal and OpenAI-family-oriented. The proper cross-provider normalization belongs in core and is tracked in #116. Note the client usage UI now performs the same uncached derivation in `useChat`, so both surfaces agree until #116 lands a normalized `Usage` API. Logging raw prompt/cache values only is a fine alternative if you'd prefer that interim — happy to switch.

File: samples/LmStreaming.Sample/WebSocket/ChatWebSocketManager.cs (hunk 5 above)
Comment (gautam-achieveai) [reply samples/LmStreaming.Sample/WebSocket/ChatWebSocketManager.cs:232]: Answer: today `Usage.InputTokenDetails.CachedTokens` preserves **provider-native** semantics, which differ by provider:
- **OpenAI family (Responses + chat completions):** `cached_tokens` is a **subset of** `PromptTokens` → uncached = `PromptTokens - CachedTokens`.
- **Anthropic:** the parser sets `PromptTokens` to the *uncached* input and `CachedTokens` to the **additive** `cache_read_input_tokens` (cached is NOT part of prompt).

No field distinguishes the two, which is exactly the gap #116 will close with a normalized uncached-input contract/helper. Until then consumers must be provider-aware; the sample log + client UI use the `cached <= prompt` guard as a pragmatic stand-in.

**Comment by gautam-achieveai** [comment]
Thanks for the reviews 🙏 Addressed the actionable findings in `eda44db`:

- ✅ **MEDIUM — streaming coverage:** added a streaming Chat Completions usage test driving `OpenClientAgent.GenerateReplyStreamingAsync` (asserts cached/reasoning survive `OpenMessage.ToStreamingMessage`).
- ✅ **MEDIUM — non-throwing parse:** `TryReadNestedInt` now uses `TryGetInt32`.
- ✅ **LOW — duplication:** centralized the cached/reasoning → core mapping behind `Usage.WithTokenDetails(...)`, reused in all three sites.
- ↪️ **LOW — sample-log provider semantics** + ❓ **QUESTION — `CachedTokens` contract:** replied inline. Net: `CachedTokens` is currently provider-native (subset of prompt for OpenAI, additive for Anthropic); the normalized cross-provider contract is tracked in #116.
- 📌 **`OpenUsage.IsCached` redundancy** (Gautam's bot): fair call — left as a non-blocking cleanup to keep this PR scoped; noted for a follow-up.

Also folded in a related fix surfaced by real run telemetry: the client usage banner's **"In" was showing total input (double-counting Cached)**. It now shows uncached input (`prompt − cached`, guarded), so **In + Cached + Out reconciles to Total**. Covered by `useChatUsage.test.ts` (single-turn, multi-turn accumulation, additive-cache fallback).

All affected C# + client suites green; full solution build clean.

**Review by gautam-achieveai — COMMENTED** [review]
**[Gautam's review bot]** :white_check_mark: **APPROVE**

**Incremental review** of `eda44db` - all 5 previous review comments have been cleanly addressed:

| Feedback | Resolution |
|----------|-----------|
| MEDIUM: Missing streaming test | Added `Streaming_UsageMessage_preserves_cached_and_reasoning_tokens_from_chat_completions` |
| MEDIUM: `GetInt32()` can throw | Switched to `TryGetInt32()` in `TryReadNestedInt` |
| LOW: 3-way token-detail duplication | Centralized in `Usage.WithTokenDetails(cached, reasoning)` |
| LOW: Provider semantics in sample layer | Acknowledged, tracked in #116 |
| QUESTION: CachedTokens contract | Answered - provider-native semantics today |

**New work in this iteration:**
- `Usage.WithTokenDetails()` is a clean addition to LmCore - proper record-with pattern, idempotent on non-positive counts
- Frontend `uncachedInputTokens` accounting is sound: In + Cached + Out == Total holds for OpenAI family, defensive fallback prevents negatives for Anthropic-style providers
- New `useChatUsage.test.ts` covers single-turn, multi-turn accumulation, and the additive-cache edge case

All relevant .NET tests pass (3/3 targeted). No blocking issues found.
